CONFIG.md
<!-- NOTE Right now this only lists the most important options
Some of these may be outdated or not be supported by iCEFARM.
Currently, reading annotations of data/farmconfig.ini in addition to these
values is recommended.
TODO not really sure what to do with these right now
moved temporarily so that the readme is less cluttered to make install easier
probably want to put in website eventually instead
-->
### Contents
- [Simulation Modes](#simulation-modes)
- [Fitness Function Parameters](#fitness-function-parameters)
- [GA parameters](#ga-parameters)
- [Initialization Parameters](#initialization-parameters)
- [Stopping Conditions Parameters](#stopping-conditions)
- [Logging parameters](#logging-parameters)
- [System parameters](#system-parameters)
- [Hardware parameters](#hardware-parameters)

The `TOP-LEVEL PARAMETER` `base_config` can be modified to specify a base config file.
In the event of a missing config parameter, it will be filled in with that from the base config file.
These can be stacked into a line of configuration files.
The configuration files will be combined and the resulting file to use for evolution will
be output. All parameters still must be specified at some point in the final configuration.

#### Simulation Modes
| Mode | Description |
|--------|-------------|
| FULLY_INTRINSIC | Compiles hardware files, then uploads to FPGA |
| REMOTE | Same as FULLY_INTRINSIC, but evaluates circuits remotely using iCEFARM system after compilation |
| SIM_HARDWARE | Compiles hardware files, but arbitrarily evaluates fitness all on the host computer. This option is useful for verifying that mutation and crossover are functional |
| FULLY_SIM | Generates a bitstream that represents a sum of sine waves. This is treated as a waveform generated by a variance measure |


*Note: FULLY_INTRINSIC or REMOTE should be used unless verification of some process is being done

#### Fitness Function Parameters
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Fitness function | The fitness function to use | TOLERANT_PULSE_COUNT, SENSITIVE_PULSE_COUNT, VARIANCE, COMBINED |  |
| Desired frequency | If using the pulse fitness function, the target frequency of the evolved oscillator | (In Hertz) 1 - 1000000 | 1000 |
| COMBINED_MODE | If using the combined fitness function, how to combine the fitnesses | ADD, MULT | |
| PULSE_WEIGHT | If using the combined fitness function, what weigthing to use for closeness to the trigger voltage in combined fitness| 0.0 - 1.0 | |
| VAR_WEIGHT | If using the combined fitness function, what weigthing to use for variance in combined fitness | 0.0 - 1.0 | |
| NUM_SAMPLES | Number of samples to record in pulse count fitness functions. The minimum number recorded will be used to determine the actual pulse fitness. Higher number of samples will take longer to
run, but should result in more stable circuits | 1+ | 1-5 |

#### GA parameters
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Population size | The number of circuits to evolve | 2 - 1000+ | 10 - 50 |
| Mutation type | Mutation algorithm to use | Simple, Rank, Proportional, Convergence | Rank |
| Mutation probability | The probability to flip a bit of the bitstream during mutation | 0.0 - 1.0 | (1 / genotypic length) = 0.0021 |
| Crossover probability | The probability of replacing a bit in one bitstream from a bit from another during crossover | 0.0 - 1.0 | 0.1 - 0.5 |
| Elitism fraction | The percentage of most fit circuits to protect from modification in a given generation | 0.0 - 1.0 | 0.1 |
| Selection | The type of selection to perform | SINGLE_ELITE, FRAC_ELITE, CLASSIC_TOURN, FIT_PROP_SEL, RANK_PROP_SEL | FIT_PROP_SEL |
| Diversity measure | The method to use to measure diversity | NONE, UNIQUE, HAMMING_DIST | HAMMING_DIST |
| Random injection | Thr probability of randomly injecting circuits into each generation | 0.0 - 1.0 | 0.0 - 0.15 |
| Chaos injection | Randomly performs additional mutations on 10% of circuits after 5 generations without fitness increase. Mutations are done at mutation_chance * chaos_injection | 0.0+ | 5 |

##### Selection methods
| Method | Description |
|--------|-------------|
| SINGLE_ELITE | Mutates the hardware of every circuit that is not the current best circuit |
| FRAC_ELITE | Creates a group of elite circuits from the population whose size is based on the elitism percentage. Mutates the hardware of and performs crossover on every non-elite |
| CLASSIC_TOURN | Randomly pairs together every circuit in the population and compares them. Keeps the winner the same and mutates and performs crossover on the loser |
| FIT_PROP_SEL | Creates a group of elite circuits from the population whose size is based on the elitism percentage. Every non-elite is compared to a random elite chosen based on the elites' fitnesses and is mutated and crossed with the elite |
| RANK_PROP_SEL | Same as above, but the elite is chosen randomly based on the elites' ranks |
| MAP_ELITES | MAP Elites-inspired selection method, for variance experiments. Maps circuits based on min/max voltage into 50x50 cells. The top individuals in each cell are copied and mutated for the next generation. |

#### Initialization Parameters
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Init mode | The method to generate the initial random circuits | CLONE_SEED, CLONE_SEED_MUTATE, RANDOM, EXISTING_POPULATION | RANDOM, CLONE_SEED_MUTATE |
| Randomize until | The method used for randomizing the initial population | PULSE, VARIANCE, NO | NO |
| Randomize threshold | The target fitness for initial random search before evolution begins| 3-8 | 4 |
| Randomize mode | The method to use when "randomizing" each circuit | MUTATE, RANDOM | Depends on the situation. If a seed individual/population is used, then use MUTATE. Otherwise, use RANDOM |
| Seed | Circuit path to use as seed | Any .asc filepath | data/seed-hardware.asc |

##### Initialization Modes
| Mode | Description |
|--------|-------------|
| CLONE_SEED | Clones the seed hardware to every individual in the population (i.e. all individuals are the same at the start) |
| CLONE_SEED_MUTATE | Clones the seed hardware to every individual in the population, and mutates each individual |
| RANDOM | Randomly assigns all (modifiable) bits of each individual |
| EXISTING_POPULATION | Completely copies the existing population from the specified directory |

*Note: The FULLY_SIM simulation mode will use the RANDOM initialization mode every time

#### Stopping Conditions
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Generations | The maximum number of generations to iterate through | 2 - 1000+ or IGNORE | 50 - 500 |
| Target Fitness | The goal fitness; evolution terminates once any individual reaches this | 1-1000+ or IGNORE | IGNORE |

#### Logging parameters
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Log Level | The amount of logs to show; higher log level means more detailed logs are shown | 1-4 | 2 |
| Save Log | Wether or not to save the logging output in a file | true, false | true |
| Save Plots | Wether or not to save the plots as images throughout evolution | true, false | true |
| Log Scale Pulses | Use symmetric log scale for pulse count y-axes | true, false | false |
| Log Scale Fitness | Use symmetric log scale for the fitness y-axis, with a reference line at fitness=1.0 | true, false | false |
| Backup Workspace | Wether or not to save the workspace directory in a backup folder after evolution | true, false | true |
| Log File | The file to save log output in | Any file path | ./workspace/log |
| Plots Directory | The directory to put the plots in | Any directory | ./workspace/plots |
| Output Directory | The directory to store previous workspaces in | Any directory not in ./workspace | ./prev_workspaces |
| ASC Directory | The directory to put the asc files (raw bitstreams) | Any directory | ./workspace/experiment_asc |
| BIN Directory | The directory to put the bin files (compiled bitstreams) | Any directory | ./workspace/experiment_bin |
| Data Directory | The directory to put the data files (MCU read data) | Any directory | ./workspace/experiment_data |
| Analysis Directory | The directory to put the analysis files | Any directory | ./workspace/analysis || Best file | The path to put the asc file of the best performing circuit throughout evolution | Any file path | ./workspace/best.asc |
| Source Populations Directory | The directory consisting of source populations to use in initialization | Any directory | ./workspace/source_populations |
| Generations Directory | The directory to put generation files into, when populations are saved each generation. The reconstruct command pulls from this directory | Any directory | ./workspace/generations |
| Use Overall Best | Whether or not to draw the overall best line in the plots | true or false | true |

#### System parameters
| Parameter | Description | Possible Values |
|-----------|-------------|-----------------|
| USB Path | The path to the USB device file | Any device file path (e.g. `/dev/ttyUSB0`) |

#### Hardware parameters
| Parameter | Description | Possible Values | Recommended Values |
|-----------|-------------|-----------------|--------------------|
| Routing | Specifies what surround tiles a logic tile can connect to | MOORE, NEWSE | MOORE |
| MCU Read Timeout | How long to wait to read from the mcu | 1+ | 1.1|
| Serial Buad | The baudrate to use for serial communication | 300, 600, 1200, 2400, 4800, 9600, 14400, 19200, 28800, 31250, 38400, 57600, and 115200 | 115200 |
| Accessed Columns | The columns in each logic tile's bitstream to modify throughout evolution | List of comma seperated numbers from 0 to 53 | 14,15,24,25,40,41|

#### iCEFARM parameters
| Parameter | Description | Possible Values |
|-----------|-------------|-----------------|
| Url | Url to iCEFARM control server | Any Url |
| Mode | How to distribute evaluations across remote devices | quick, all |
| Devices | Amount of devices to reserve from iCEFARM | 1+ |

dev/_sources/code/PlotConfig.rst
=============
PlotConfig.py
=============
.. automodule:: PlotConfig
    :members:
    :private-members:

config.py
# Bitstream Evolution Config File
# This is where you setup an experiment to run.

# Set BitstreamEvolution (a.k.a. The folder this config is in) as the current working directory
import os
from pathlib import Path
from random import Random

from Directories import Directories
from EvaluateFitness.EvaluateFitness import EvaluateFitness
from Evolution import Evolution
from GenerateMeasurements.GenerateMeasurements import SimpleGenerateMeasurements
from Hardware.Microcontroller import Microcontroller, MicrocontrollerConfig
from Logger import Logger, LoggerConfig
from PlotConfig import save_plot_config
from PlotDataRecorder import PlotDataRecorder
from Population.PopulationInitialization import FileBasedCircuitFactory, GenerateBitstreamPopulation, GenerateSinglePopulationWrapper, NoPostConstructionStrategy, NoRandomizationStrategy
os.chdir(os.path.dirname(os.path.abspath(__file__)))

from BitstreamEvolutionProtocols import DataRequest, GenDataIncrementer
from EvaluateFitness.EvalVarMaxFitness import EvalVarMaxFitness

explanation = '<empty explanation>'

run_fitness_sensitivity_exp = False
population_size = 50
num_generations = 200
bitstream_length = 1728
fpga = 'i:0x0403:0x6010:0'
post_construction_strategy = NoPostConstructionStrategy()
randomization_strategy = NoRandomizationStrategy()
mutation_prob = 0.0021
crossover_prob = 0.7
uses_init_existing_population = False
num_samples = 1

plot_data_recorder = PlotDataRecorder()

desired_frequency = 0
data_request = DataRequest.WAVEFORM
evaluation = EvalVarMaxFitness(plot_data_recorder)

log_file = Path('workspace', 'log')
log_level = 0
save_log = False

plots_dir = Path('workspace', 'plots')
launch_plots = True
save_plots = True
frame_interval = 100

usb_path = '/dev/ttyUSB0'
serial_baud = 115200
read_timeout = 1.1
routing_type = 'MOORE'
accessed_columns = [14, 15, 24, 25, 40, 41]

asc_dir=Path('workspace', 'experiment_asc')
bin_dir=Path('workspace', 'experiment_bin')
data_dir=Path('workspace', 'experiment_data')

is_pulse_count = False
is_tone_discriminator = False

transfer_interval = 0
diversity_measure = 'HAMMING_DIST'

########################################################

save_plot_config(Path('workspace', 'plot_config.ini'), 
                 pop_sz=population_size,
                 is_pulse_count=is_pulse_count,
                 is_td=is_tone_discriminator,
                 desired_freq=desired_frequency,
                 transfer_interval=transfer_interval,
                 diversity_measure=diversity_measure,
                 save_plots=save_plots,
                 plots_dir=str(plots_dir),
                 uses_init_existing_population=uses_init_existing_population)

logger_config = LoggerConfig(
    plots_dir=plots_dir,
    launch_plots=launch_plots,
    is_sensitivity=run_fitness_sensitivity_exp,
    frame_interval=frame_interval,
    log_file=log_file,
    log_level=log_level,
    save_log=save_log,
    datetime_format='%%m/%%d/%%Y - %%H:%%M:%%S'
)

mcu_config = MicrocontrollerConfig(
    usb_path=usb_path,
    serial_baud=serial_baud,
    read_timeout=read_timeout
)

directories = Directories(
    asc_dir,
    bin_dir,
    data_dir
)

rand = Random()

logger = Logger(explanation, logger_config)

evolution = Evolution(
    gen_data_factory=GenDataIncrementer(num_generations),

    circuit_factory=FileBasedCircuitFactory(
        sz=population_size,
        logger=logger,
        directories=directories,
        routing_type=routing_type,
        accessed_columns=accessed_columns
    ).create,

    reproduce=None,

    gen_init_populations=GenerateSinglePopulationWrapper(
        GenerateBitstreamPopulation(
            sz=population_size,
            bitstream_sz=bitstream_length,
            post_construction_strategy=post_construction_strategy,
            randomization_strategy=randomization_strategy,
            mutation_prob=mutation_prob,
            rand=rand
        ).generate
    ).generate,

    eval_population_fitness=EvaluateFitness(evaluation).evaluate,

    generate_measurements=SimpleGenerateMeasurements(fpga, data_request, num_samples).generate,

    hardware=Microcontroller(
        fpga=fpga,
        logger=logger,
        config=mcu_config
    ),

    plot_data_recorder=plot_data_recorder
)
evolution.run()

